oauth2.go

```diff
@@ -224,6 +226,64 @@ func (c *Config) Exchange(ctx context.Context, code string, opts ...AuthCodeOpti
 	return retrieveToken(ctx, c, v)
 }
 
+// AuthDevice returns a device auth struct which contains a device code
+// and authorization information provided for users to enter on another device.
+func (c *Config) AuthDevice(ctx context.Context, opts ...AuthCodeOption) (*DeviceAuth, error) {
+	v := url.Values{
+		"client_id": {c.ClientID},
+	}
+	if len(c.Scopes) > 0 {
+		v.Set("scope", strings.Join(c.Scopes, " "))
+	}
+	for _, opt := range opts {
+		opt.setValue(v)
+	}
+	return retrieveDeviceAuth(ctx, c, v)
+}
+
+// Poll does a polling to exchange an device code for a token.
+func (c *Config) Poll(ctx context.Context, da *DeviceAuth, opts ...AuthCodeOption) (*Token, error) {
+	v := url.Values{
+		"client_id":   {c.ClientID},
+		"grant_type":  {"urn:ietf:params:oauth:grant-type:device_code"},
+		"device_code": {da.DeviceCode},
+	}
+	if len(c.Scopes) > 0 {
+		v.Set("scope", strings.Join(c.Scopes, " "))
+	}
+	for _, opt := range opts {
+		opt.setValue(v)
+	}
+
+	// If no interval was provided, the client MUST use a reasonable default polling interval.
+	// See https://tools.ietf.org/html/draft-ietf-oauth-device-flow-07#section-3.5
+	interval := da.Interval
+	if interval == 0 {
+		interval = 5
+	}
+
+	for {
+		time.Sleep(time.Duration(interval) * time.Second)
+
+		tok, err := retrieveToken(ctx, c, v)
+		if err == nil {
+			return tok, nil
+		}
+
+		errTyp := parseError(err)
+		switch errTyp {
+		case errSlowDown:
+			interval += 5
+		case errAuthorizationPending:
+			// Do nothing.
+		case errAccessDenied, errExpiredToken:
+			fallthrough
+		default:
+			return tok, err
+		}
+	}
+}
```

Comment:
Using a ticker and `select` allows for handling the context timeout:

```suggestion
        ticker := time.NewTicker(time.Duration(interval) * time.Second)
        for {
                select {
		case <-ctx.Done():
                        return nil, ctx.Err()
	        case <-ticker.C:
                        tok, err := retrieveToken(ctx, c, v)
                        if err == nil {
                                return tok, nil
                        }

                        errTyp := parseError(err)
                        switch errTyp {
                        case errSlowDown:
                                interval += 5
                                ticker.Reset(time.Duration(interval) * time.Second)
                                continue
                        case errAuthorizationPending:
                                // Do nothing.
                                continue
                        case errAccessDenied, errExpiredToken:
                                fallthrough
                        default:
                                return tok, err
			}
		}
	}
}
```

Reply:
Done.